### PR TITLE
[FSTORE-2048] Improve monitoring on feature embeddings and model re-training automation

### PR DESCRIPTION
**Pull request by javierdlrm**
## Summary
Extends Feature Monitoring V2 to features of type **embedding** (vectors), which were previously skipped by statistics computation. Spark/JVM engine only; Python-engine profiling parity, per-dimension stats, and model re-training automation are out of scope (the latter tracked separately).

- **Spark/JVM profiler**: numeric `array<T>` columns are profiled as `Embedding`. Per-row L2 norm runs through the existing histogram + KLL path, and the element-wise mean (centroid) is computed, both emitted under a new `embedding` block in the profile JSON. Row validity rules: null / empty / null-NaN-or-infinite element / ragged length / all-invalid.
- **Python SDK**: parses the embedding block; accepts embedding features at config time for distribution metrics and the new `CENTROID_DISTANCE` scalar metric (rejecting scalar metrics on them); sources norm histogram/KLL from the embedding block; merges per-batch norm KLL and a count-weighted centroid for rolling reference windows; scores `CENTROID_DISTANCE` as plain L2 distance; defaults embeddings to equi-frequency binning; reprofiles stale windows missing the embedding block.
- **Java SDK**: preserves the `embedding` block in the profiler-JSON parser.

Spec: https://hopsworks.atlassian.net/wiki/spaces/~597211656/pages/1533771777

## Test plan
- [x] `java/spark` profile tests incl. new `ColumnProfilerEmbeddingTest` (norm, centroid, edge cases, infinity)
- [x] Python `core/` tests: FDS parsing, config validation, distribution sourcing, rolling-reference merge, centroid L2, binning default, stale-stats reprofile guard
- [x] Java SDK FDS round-trip test preserves embedding
- [ ] Cluster integration: full Spark profile → HopsFS extended-statistics round-trip → drift result on a real embedding FG (deferred; no cluster yet)

## Notes for reviewers
- `CENTROID_DISTANCE` is plain L2 for the MVP; similarity-function-aware (cosine/dot) distance is a deferred follow-up.
- Companion PRs: hopsworks-ee (backend enum + validation), hopsworks-front (timeline rendering).

### CODE REVIEW

File: python/hsfs/core/distribution_engine.py [hunk 1]
```diff
@@ -226,6 +231,79 @@ def _resolve_merged_reference(
...
+        embedding_block: dict = {"norm": norm_block}
+        if merged_centroid is not None:
+            embedding_block["centroid"] = merged_centroid
+
+        return FeatureDescriptiveStatistics(
+            feature_name=first_fds.feature_name,
+            feature_type=first_fds.feature_type,
+            count=parsed.get("n"),
+            extended_statistics={"embedding": embedding_block},
+        )
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsfs/core/distribution_engine.py:327]: Fixed in f533ad64 — the merged embedding reference now carries `dimension` (centroid length, with a constituent-batch fallback) and `count`, matching the profiler-emitted block.

File: java/spark/src/test/java/com/logicalclocks/hsfs/spark/engine/profile/ColumnProfilerEmbeddingTest.java [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+        "rows with an infinite element must be dropped");
+    Assertions.assertEquals(3.0, embedding.get("centroid").get(0).asDouble(), 1e-9);
+    Assertions.assertEquals(4.0, embedding.get("centroid").get(1).asDouble(), 1e-9);
+
+    // (b) the norm histogram and KLL stay finite (an infinite norm would poison both).
+    JsonNode norm = embedding.get("norm");
+    for (JsonNode entry : norm.get("histogram")) {
+      double value = entry.get("value").asDouble();
+      Assertions.assertTrue(Double.isFinite(value), "norm histogram bin value must be finite");
+    }
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply java/spark/src/test/java/com/logicalclocks/hsfs/spark/engine/profile/ColumnProfilerEmbeddingTest.java]: Fixed in f533ad64 — the assertion now checks the numeric KLL bucket bounds and the sketch min/max (`Double.isFinite`) instead of the string histogram label, so it genuinely fails if an infinite norm leaks in.

File: python/hsfs/core/feature_monitoring_config_engine.py [hunk 3]
```diff
@@ -562,6 +581,13 @@ def _run_feature_monitoring(
                 else None
             )
 
+        # Features whose comparison config targets an embedding metric (a distribution
+        # metric, computed over the embedding norm histogram, or the CENTROID_DISTANCE
+        # scalar metric). Reused registered statistics for these features must carry the
+        # embedding block; otherwise the window is reprofiled (see
+        # MonitoringWindowConfigEngine._run_single_window_monitoring).
+        embedding_feature_names = self._get_embedding_targeted_feature_names(config)
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsfs/core/feature_monitoring_config_engine.py]: Fixed in f533ad64 — corrected the comment: only CENTROID_DISTANCE-targeted features are forwarded; distribution-metric embeddings are caught downstream via the reused FDS self-identifying as `feature_type == "Embedding"`.

File: python/hsfs/core/monitoring_window_config_engine.py [hunk 4]
```diff
@@ -311,6 +312,12 @@ def _run_single_window_monitoring(
                 read path filters rows by ``model_name = X AND model_version = str(Y)``
                 and the precomputed-stats lookup is skipped (registered stats are
                 aggregated over the whole entity and are not model-aware).
+            embedding_feature_names: optional names of features whose monitoring config
+                targets embedding metrics (a distribution metric or CENTROID_DISTANCE).
+                When registered statistics are reused for the window, any such feature
+                whose reused statistics predate embedding support (no
+                ``extended_statistics.embedding`` block) forces a recompute of the whole
+                window so the embedding statistics are present.
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsfs/core/monitoring_window_config_engine.py:324]: Fixed in f533ad64 — docstring corrected to reflect that only CENTROID_DISTANCE features are forwarded.

File: python/hsfs/core/feature_descriptive_statistics.py [hunk 5]
```diff
@@ -229,7 +235,7 @@ def id(self) -> int | None:
     @public
     @property
     def feature_type(self) -> str:
-        """Data type of the feature. It can be one of Boolean, Fractional, Integral, or String."""
+        """Data type of the feature. It can be one of Boolean, Fractional, Integral, String, or Embedding."""
         return self._feature_type
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsfs/core/feature_descriptive_statistics.py:242]: Fixed in f533ad64 — split into one sentence per line.

File: python/hsfs/core/feature_monitoring_result_engine.py [hunk 6]
```diff
@@ -744,8 +792,12 @@ def _default_binning_strategy(fds: FeatureDescriptiveStatistics) -> str:
     """Return the default binning strategy for the given feature statistics.
 
     Uses CATEGORICAL for non-numeric features (string, boolean, etc.)
-    and EQUI_FREQUENCY for numeric ones.
+    and EQUI_FREQUENCY for numeric ones. Embeddings bin over their numeric norm
+    histogram, so they use EQUI_FREQUENCY too.
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsfs/core/feature_monitoring_result_engine.py]: Fixed in f533ad64 — split into one sentence per line.

File: python/hsfs/core/monitoring_window_config_engine.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsfs/core/monitoring_window_config_engine.py:324]: Fixed in f533ad64 — docstring corrected (CENTROID_DISTANCE only; distribution metrics excluded).

File: python/hsfs/core/monitoring_window_config_engine.py [hunk 7]
```diff
@@ -394,6 +401,20 @@ def _run_single_window_monitoring(
                 row_percentage=monitoring_window_config.row_percentage,
             )
 
+        # Stale-stats guard: registered statistics may predate embedding support and so
+        # lack the ``extended_statistics.embedding`` block. Reusing them for an embedding
+        # feature would silently drop the configured embedding metric or trip the
+        # detection/reference subset assertion downstream. Drop the reused stats in that
+        # case so the recompute path below reprofiles the window with embedding support.
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsfs/core/monitoring_window_config_engine.py]: Fixed in f533ad64 — split into one sentence per line.

File: python/hsfs/core/feature_monitoring_config_engine.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsfs/core/feature_monitoring_config_engine.py]: Fixed in f533ad64 — comment corrected to match the implementation.

File: python/hsfs/core/feature_monitoring_result_engine.py [hunk 8]
```diff
@@ -507,6 +516,45 @@ def _compute_distribution_distance(
...
+        detection_statistics: FeatureDescriptiveStatistics,
+        reference_statistics: FeatureDescriptiveStatistics | None,
+    ) -> float | None:
+        """Compute the L2 distance between the detection and reference centroids.
+
+        This is the MVP centroid drift score and always uses the plain Euclidean
+        distance. Threading a configurable similarity function is a deferred
+        follow-up. Returns None when either centroid is missing, when either window
+        is empty, or when the centroid lengths do not match.
+        """
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsfs/core/feature_monitoring_result_engine.py:533]: Fixed in f533ad64 — split into one sentence per line.

File: python/hsfs/core/feature_monitoring_result_engine.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsfs/core/feature_monitoring_result_engine.py]: Fixed in f533ad64 — split into one sentence per line.

File: java/spark/src/main/java/com/logicalclocks/hsfs/spark/engine/profile/ColumnProfiler.java [hunk 9]
```diff
@@ -372,6 +423,144 @@ private void buildNumericFields(Dataset<Row> df, ColumnInfo col, Row aggRow, lon
...
+   * <p>Uses {@code posexplode} to map each vector to {@code (pos, value)} rows, averages per
+   * position, and collects ordered by position into a dense {@code double[]} of length
+   * {@code dimension}.
+   */
+  private double[] computeCentroid(Dataset<Row> validRows, String columnName, int dimension) {
+    String quoted = "`" + columnName.replace("`", "``") + "`";
+    Dataset<Row> exploded = validRows
+        .select(functions.posexplode(functions.col(columnName)))
+        .groupBy(functions.col("pos"))
+        .agg(functions.avg(functions.col("col").cast("double")).alias("_avg"));
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply java/spark/src/main/java/com/logicalclocks/hsfs/spark/engine/profile/ColumnProfiler.java:551]: Fixed in f533ad64 — removed the unused `quoted` local in `computeCentroid`.

File: python/hsfs/core/distribution_engine.py [hunk 10]
```diff
@@ -226,6 +231,90 @@ def _resolve_merged_reference(
...
+            constituent_dimension = _first_embedding_dimension(reference_fds_list)
+            if constituent_dimension is not None:
+                embedding_block["dimension"] = constituent_dimension
+
+        return FeatureDescriptiveStatistics(
+            feature_name=first_fds.feature_name,
+            feature_type=first_fds.feature_type,
+            count=merged_count,
+            extended_statistics={"embedding": embedding_block},
+        )
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsfs/core/distribution_engine.py:327]: Fixed in ac7c6d5c — the merged embedding reference now sets top-level `percentiles`/`min`/`max` from the KLL-merger output, mirroring the scalar merged path, so embedding equi-frequency binning no longer falls back to equi-width.

File: python/hsfs/core/distribution_engine.py [hunk 11]
```diff
@@ -408,7 +511,7 @@ def _get_extended_stats(
     ) -> dict | None:
         cache_key = (fds.id, window_id)
         if cache_key not in self._cache:
-            self._cache[cache_key] = fds.extended_statistics or {}
+            self._cache[cache_key] = _extended_statistics_for(fds)
         return self._cache[cache_key]
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsfs/core/distribution_engine.py:517]: Fixed in b9d4c213 — the cache key now includes the feature name ((id, feature_name, window_id)), so synthetic merged FDS with id=None no longer collide across features.

File: python/hsfs/core/monitoring_window_config_engine.py [hunk 12]
```diff
@@ -620,6 +646,40 @@ def _round_and_convert_event_time(self, event_time: datetime) -> int | None:
...
+        """
+        fds_list = registered_stats.feature_descriptive_statistics
+        if not fds_list:
+            return False
+        targeted = embedding_feature_names or set()
+        for fds in fds_list:
+            requires_embedding = fds.feature_name in targeted or fds._is_embedding()
+            if requires_embedding and fds._get_embedding_statistics() is None:
+                return True
+        return False
```
[the remark this answers is not in the document]
Comment (javierdlrm) [reply python/hsfs/core/monitoring_window_config_engine.py]: Fixed in d83aa1f0 — `_reused_stats_lack_embedding` now returns True when a CENTROID_DISTANCE-targeted feature is absent from the reused stats entirely (not just present-but-missing-block), forcing the reprofile.
